var/spack/repos/builtin/packages/strumpack/package.py

```diff
@@ -148,44 +148,55 @@ def cmake_args(self):
                 args.append('-DHIP_HIPCC_FLAGS=--amdgpu-target={0}'.
                             format(",".join(rocm_archs)))
 
-        if self.spec.satisfies('@:5.1.1'):
-            self.test_data_dir = 'examples/data'
-        else:
-            self.test_data_dir = 'examples/sparse/data'
-        self.test_src_dir = 'test'
         return args
 
+    test_src_dir = 'test'
+
+    @property
+    def test_data_dir(self):
+        """Return the stand-alone test data directory."""
+        add_sparse = not self.spec.satisfies('@:5.1.1')
+        return join_path('examples', 'sparse' if add_sparse else '', 'data')
+
     @run_after('install')
     def cache_test_sources(self):
         """Copy the example source files after the package is installed to an
         install test subdirectory for use during `spack test run`."""
         self.cache_extra_test_sources([self.test_data_dir, self.test_src_dir])
 
     def _test_example(self, test_prog, test_dir, test_cmd, test_args):
-        tmpbld_dir = '{0}/_BUILD'.format(test_dir)
-        with working_dir(tmpbld_dir, create=True):
-            with open('{0}/CMakeLists.txt'.format(tmpbld_dir), 'w') as mkfile:
-                mkfile.write('cmake_minimum_required(VERSION 3.13)\n')
-                mkfile.write('project(StrumpackSmokeTest LANGUAGES CXX)\n')
-                mkfile.write('find_package(STRUMPACK REQUIRED)\n')
-                mkfile.write('add_executable({0} ../{0}.cpp)\n'.
-                             format(test_prog))
-                mkfile.write('target_link_libraries({0} '.format(test_prog) +
-                             'PRIVATE STRUMPACK::strumpack)\n')
-
-            opts = self.std_cmake_args
-            opts += self.cmake_args()
-            opts += ['.']
-            self.run_test('cmake', opts, [], installed=False, work_dir='.')
-            self.run_test('make')
-            with set_env(OMP_NUM_THREADS='4'):
-                self.run_test(test_cmd, test_args, installed=False,
-                              purpose='test: strumpack smoke test',
-                              skip_missing=False, work_dir='.')
-        self.run_test('rm', ['-fR', tmpbld_dir])
+        if not self.spec.satisfies('+openmp+parmetis'):
+            print('Skipping {0} test: +openmp+parmetis required'.
```

Comment:
Looking at the code, it seems that METIS (not PARMETIS) is the default for both test_sparse_seq and test_sparse_mpi, so I would think parmetis is not required. Do you get errors if parmetis or openmp are not enabled?

Reply:
Yes, I do get errors when I don't have them installed. Building the test fails in the `CMake` stage due to METIS-related errors.  I can paste the output when I reproduce it.  

I chose `parmetis` since it is a variant.  

I'm not so confident about `OpenMPI` except that the test does set `OMP_NUM_THREADS` before running the program.

Reply:
For `test_sparse_mpi` you can enable parmetis (instead of metis) by running as
`mpirun -n 4 ./test_sparse_mpi ../../examples/sparse/data/pde900.mtx --sp_reordering_method parmetis`
If not compiled with `+parmetis` this test will fail.

Reply:
You mean `OpenMP`? It should still work correctly if not compiled with `+openmp`, even if `OMP_NUM_THREADS` are set.

Reply:
> You mean `OpenMP`? 
🤦 Yes.

>It should still work correctly if not compiled with `+openmp`, even if `OMP_NUM_THREADS` are set.

Sure.

Are there any other examples or tests that you'd recommend that could be compiled and run if those variants aren't aren't enabled?  Or are `metis`/`parmetis` required?

Reply:
`metis` is always requires, but `parmetis` is optional.
For the MPI test, you can run with `parmetis` by adding the command line option `--sp_reordering_method parmetis` to the test, otherwise `metis` will be used.

Reply:
> `metis` is always requires, but `parmetis` is optional.
> For the MPI test, you can run with `parmetis` by adding the command line option `--sp_reordering_method parmetis` to the test, otherwise `metis` will be used.

That assumes the test actually builds.  Without changing the current `CMake` arguments, I get:

```
CMake Error at /usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-skylake_avx512/gcc-8.3.1/strumpack-5.1.1-ewsohjoqqwvgwvgzyqfpovk5rb4xgwoa/lib/cmake/STRUMPACK/strumpack-config.cmake:4 (list):
  list does not recognize sub-command PREPEND
Call Stack (most recent call first):
  CMakeLists.txt:3 (find_package)


CMake Error at /usr/tce/packages/cmake/cmake-3.14.5/share/cmake/Modules/CMakeFindDependencyMacro.cmake:48 (find_package):
  By not providing "FindMETIS.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "METIS", but
  CMake did not find one.

  Could not find a package configuration file provided by "METIS" with any of
  the following names:

    METISConfig.cmake
    metis-config.cmake

  Add the installation prefix of "METIS" to CMAKE_PREFIX_PATH or set
  "METIS_DIR" to a directory containing one of the above files.  If "METIS"
  provides a separate development package or SDK, be sure it has been
  installed.
```

for an installation where `~mpi~openmp~parmetis` and the installation with `+mpi~openmp` seems to hang.

Reply:
Adding `CMAKE_MODULE_PATH` to point to the installation `cmake` subdirectory, yields:

```
...
-- Found METIS: /usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-skylake_avx512/gcc-8.3.1/metis-5.1.0-n3sspawwpesqahok5yket7wmmpnjunat/lib/libmetis.so (found version "5.1.0")  
-- Found ZFP: /usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-skylake_avx512/gcc-8.3.1/zfp-0.5.5-c4fi3elruxogdjbemacyao52okk5llyf/lib64/cmake/zfp/zfp-config.cmake (found version "0.5.5") 
-- Configuring incomplete, errors occurred!
See also "/g/g21/dahlgren/.spack/test/hiusihegfj67f3efue726iorvldy7dbh/strumpack-5.1.1-ewsohjo/cache/strumpack/test/CMakeFiles/CMakeOutput.log".
CMake Error at /usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-skylake_avx512/gcc-8.3.1/strumpack-5.1.1-ewsohjoqqwvgwvgzyqfpovk5rb4xgwoa/lib/cmake/STRUMPACK/strumpack-config.cmake:4 (list):
  list does not recognize sub-command PREPEND
Call Stack (most recent call first):
  CMakeLists.txt:3 (find_package)


CMake Error at CMakeLists.txt:5 (target_link_libraries):
  Cannot specify link libraries for target "test_sparse_seq" which is not
  built by this project.
...
```

Reply:
It looks like your CMake version is relatively old.
This line
https://github.com/pghysels/STRUMPACK/blob/abbcdc2b2aad92b57bf5ee1a238d7c3dca72c718/cmake/strumpack-config.cmake.in#L4
only works with CMake 3.15 and up.
I need to update the minimum required CMake version in that file.

Reply:
Good to know. It is picking up my system crake instead of the one used to build the software. This is a known issue with the tests.

Reply:
I modified the test with a known hack to address the issue until Spack handles it properly.